schema/query/projection_evaluator.go

```diff
@@ -79,47 +197,57 @@ func evalProjection(ctx context.Context, p Projection, payload map[string]interf
 					}
 				} else if ref, ok := def.Validator.(*schema.Reference); ok {
 					// Execute sub-request in batch
-					var e Expression
-					v, isArray := val.([]interface{})
-					if isArray {
-						e = &In{Field: "id", Values: v}
-					} else {
-						e = &Equal{Field: "id", Value: val}
-					}
 					q := &Query{
 						Projection: pf.Children,
-						Predicate:  Predicate{e},
+						Predicate:  Predicate{&Equal{Field: "id", Value: val}},
 					}
 					rbr.request(ref.Path, q, func(payloads []map[string]interface{}, validator schema.Validator) error {
 						var v interface{}
-						if len(payloads) == 1 && isArray == false {
-							if payloads[0] != nil {
-								payload, err := evalProjection(ctx, pf.Children, payloads[0], validator, rbr)
-								if err != nil {
-									return fmt.Errorf("%s: error applying Projection on sub-field: %v", name, err)
-								}
-								if v, err = resolveFieldHandler(ctx, pf, def, payload); err != nil {
-									return fmt.Errorf("%s: error resolving field handler on sub-field: %v", name, err)
-								}
-							}
-						} else {
-							var err error
-							for i := range payloads {
-								if payloads[i], err = evalProjection(ctx, pf.Children, payloads[i], validator, rbr); err != nil {
-									return fmt.Errorf("%s: error applying projection on sub-field item #%d: %v", pf.Name, i, err)
-								}
+						if len(payloads) == 1 {
+							payload, err := evalProjection(ctx, pf.Children, payloads[0], validator, rbr)
+							if err != nil {
+								return fmt.Errorf("%s: error applying Projection on sub-field: %v", name, err)
 							}
-							if v, err = resolveFieldHandler(ctx, pf, def, payloads); err != nil {
+							if v, err = resolveFieldHandler(ctx, pf, def, payload); err != nil {
 								return fmt.Errorf("%s: error resolving field handler on sub-field: %v", name, err)
 							}
 						}
+						// Return `null` for missing result instead of empty object
+						if m, ok := v.(map[string]interface{}); ok && len(m) == 0 {
+							v = nil
+						}
 						resMu.Lock()
 						res[name] = v
 						resMu.Unlock()
 						return nil
 					})
+				} else if dict, ok := def.Validator.(*schema.Dict); ok {
```

Comment:
Here we could perhaps also use `FieldGetter` instead of `*schema.Dict`. Again, the `*schema.Array` check must happen before the `FieldGetter` check then, as `*schema.Array` implements `FieldGetter`.

Reply:
This is legacy reasons, explained near the top of #77

Reply:
Okay, so the latest patch will do for now.